shelley/chain-and-ledger/formal-spec/multi-sig.tex

```diff
@@ -562,68 +514,154 @@ \section{Ledger Transition for Multi-Signature}
-evalMultiSigScript (SingleSig vkh) vkhs =
-    vkh `Set.member` vkhs
+import qualified Language.PlutusTx            as P
+import           Ledger.Validation            as V
 
-evalMultiSigScript (MultiSig m ts) vkhs =
-    sum [ if evalMultiSigScript t vkhs then 1 else 0 | t <- ts ] >= m
+data MultiSig = MultiSig
+                { signatories :: [Ledger.PubKey]
+                -- ^ List of public keys of people who may sign the transaction
+                , requiredSignatures :: Integer
+                -- ^ Minimum number of signatures required to unlock
+                --   the output (should not exceed @length signatories@)
+                }
 
-verifyCases :: MultiSigScript -> [(Set VKeyHash, Bool)] -> Bool
-verifyCases script cases =
-    and [ evalMultiSigScript script keyset == expected
-        | (keyset, expected) <- cases]
+validate :: MultiSig -> () -> () -> PendingTx -> Bool
+validate multiSig@(MultiSig keys num) () () p =
+    let present = P.length (P.filter (V.txSignedBy p) keys)
+    in present `P.geq` num
 \end{verbatim}
 
-The Appendix~\ref{sec:native-multi-sign} provides some examples of how this
-scheme can be used. The intention is to have an alternative in the case for
-whenever the simple Plutus script integration would not be viable.
+The above Plutus script takes a parameter \var{multiSig} of type
+$\type{MultiSig}$ which is a list of keys $\var{keys}$ and a threshold $num$
+which indicates how many of the keys are required as signatures. When the
+validation script is called, it computes the list of keys in $keys$ which signed
+the transaction. If the length of that list is greater than or equal to $num$,
+then enough signatures for the multi-signature are present. Therefore, with
+\var{hkeys} being a value of type \type{MultiSig}, calling
+$\fun{validate}~multisig$ will return a validator script with the correct type
+according to Figure~\ref{fig:types_defs_plutus}.
+
+\subsection{Native Script Interpreter}
+\label{sec:native-script-interp}
+
+\begin{figure*}[hbt]
+  \emph{MultiSig Type}
+
+  \begin{equation*}
+    \begin{array}{rll}
+      \var{msig} & \in \ScriptMSig = & \,\,\,\type{SingleSig}~\HashKey \\
+       && \vert\,\,\type{MultSig}~Int~[\ScriptMSig]
+    \end{array}
+  \end{equation*}
+
+  \emph{Functions}
+
+  \begin{align*}
+    \fun{validateScript} & \in\ScriptMSig\to\Tx\to\Bool & \text{validate native
+                                                          script} \\
+    \fun{validateScript} & \var{msig}~\var{tx}= \\
+                         & \textrm{let}~\var{vhks}\leteq \{\fun{hashKey}~vk \vert
+                           vk \in \fun{txwitsVKey}~\var{tx}\} \\
+                         & \fun{evalMultiSigScript}~msig~vhks\\
+  \end{align*}
+  \begin{align*}
+    \fun{evalMultiSigScript} & \in\ScriptMSig\to\powerset\HashKey\to\Bool & \\
+    \fun{evalMultiSigScript} & ~\serialised{\type{SingleSig}~hk}_{hks}~\var{vhks} =  hk \in vhks \\
```

Comment:
I think we should mention somewhere above about the syntax/semantics difference between a script `s` and the function `[|s|]`, like how it is done in the "An Abstract Model of UTXO-based Cryptocurrencies with Scripts" paper at the bottom of page 3: https://eprint.iacr.org/2018/469.pdf

I think `evalMultiSigScript` _is_ `[|.|]`, and it doesn't make sense to write 
```evalMultiSigScript [|s|]```, etc.